### ISSUE TITLE
`create_bidirectional_sliding_window_mask` builds the mask on `encoder_hidden_states`' device — same model-parallel bug as #46221, which only fixed `create_bidirectional_mask`

### ISSUE THREAD

**Issue by abcgco** [huggingface/transformers#47559]
### System Info

- `transformers` main (5.15.0.dev0, commit b6d5084f)
- torch 2.12.1
- Reproduced with a cpu/mps device split; the mismatch is device-pair agnostic (any multi-device `device_map` split)

### Reproduction

#46221 fixed a model-parallel device mismatch in `create_bidirectional_mask`: `_preprocess_mask_arguments` moves the 2D `attention_mask` to `inputs_embeds.device`, but the mask was built on `embeds.device`, where `embeds = encoder_hidden_states if ... else inputs_embeds`. The sibling function `create_bidirectional_sliding_window_mask` still has the exact pre-#46221 code:

https://github.com/huggingface/transformers/blob/b6d5084fb4a5dd11e44005a5fa009e7943271090/src/transformers/masking_utils.py#L1282-L1283

```python
embeds = encoder_hidden_states if encoder_hidden_states is not None else inputs_embeds
batch_size, dtype, device = embeds.shape[0], embeds.dtype, embeds.device
```

Contrast between the fixed function and the sibling, with encoder states living on another device:

```python
import torch
from transformers import Gemma2Config
from transformers.masking_utils import create_bidirectional_mask, create_bidirectional_sliding_window_mask

# any config with a `sliding_window` attribute works; Gemma2Config used for brevity
cfg = Gemma2Config(hidden_size=32, num_attention_heads=2, num_key_value_heads=1,
                   num_hidden_layers=2, sliding_window=4, _attn_implementation="sdpa")

inputs_embeds = torch.randn(2, 5, 32, device="mps")          # decoder side
encoder_hidden_states = torch.randn(2, 7, 32, device="cpu")  # encoder side (model-parallel split)
attention_mask = torch.tensor([[1]*7, [0]*2 + [1]*5])        # padded row -> forces mask materialization

for fn in (create_bidirectional_mask, create_bidirectional_sliding_window_mask):
    mask = fn(config=cfg, inputs_embeds=inputs_embeds, attention_mask=attention_mask,
              encoder_hidden_states=encoder_hidden_states, past_key_values=None)
    print(fn.__name__, mask.device)
```

Output on main (b6d5084f):

```
create_bidirectional_mask mps:0                      # fixed by #46221
create_bidirectional_sliding_window_mask cpu         # sibling still returns the mask on the wrong device
```

The attention that consumes this mask runs on the query (`inputs_embeds`) device, so the downstream attention call hits a device mismatch — the failure mode #46221 fixed for `create_bidirectional_mask`, and the invariant its inline comment documents ("In model parallel setups, `encoder_hidden_states` may live on a different device than `inputs_embeds`").

Affected callers are the ones passing `encoder_hidden_states` to the sliding-window variant: `gemma4_assistant` and `gemma4_unified_assistant` (cross-attention to sliding-window encoder states), which can hit this under a multi-device `device_map` with a padded batch. All other call sites (`t5gemma`, `modernbert`, `openai_privacy_filter`, and the internal `is_causal=False` fallback inside `create_sliding_window_causal_mask`) never pass `encoder_hidden_states`, so both device sources coincide there.

### Expected behavior

Both bidirectional mask helpers should build the mask on `inputs_embeds.device`, consistently with `_preprocess_mask_arguments`. The fix is the same few lines as #46221, applied to the sliding-window variant.

Happy to open a PR mirroring #46221, with a regression test covering both helpers (#46221 landed without one).
